add v15+ upgrading guide (#159)
vaadin/docs: articles/guide/upgrading/v15-upgrade/overview.asciidoc

---
title: Overview
order: 0
layout: page
---

:vaadin-version: {moduleMavenVersion:com.vaadin:vaadin}
pass:[<!-- vale Vaadin.Versions = NO -->]

= Upgrading From Vaadin 15 To {vaadin-version}

> Upgrading from a Vaadin 15+ version to the latest is in general straightforward, nevertheless, a few breaking changes to take into account when upgrading.

Reviewer: ```suggestion
Upgrading from a Vaadin 15+ version to the latest is in general straightforward, however, there are some breaking changes to consider when upgrading.
```

---
title: Overview
order: 0
layout: page
---

:vaadin-version: {moduleMavenVersion:com.vaadin:vaadin}

= Upgrading From Vaadin 15 To {vaadin-version}

Upgrading from a Vaadin 15+ version to the latest is in general straightforward, nevertheless, a few breaking changes need to take into account when upgrading:

[role="since:com.vaadin:vaadin@V19"]
== Generated @Id Field Is Now Optional In TypeScript
A field with `@Id` annotation in Java is now optional in the generated TypeScript code. 
Given an entity with an `id` field like below

.Entity.java
[source,java]
----
public class Entity {
    @Id
    private int id;
}
----

now in the TypeScript files, instead of using `endpoint.getEntity(entity.id)`, you might need to change to `endpoint.getEntity(entity.id!)` or `endpoint.getEntity(entity.id?)`.

== Spring Security

The Spring Security configuration requires some changes for the service worker related files.

[role="since:com.vaadin:vaadin@V19"]

> === Ignore The `sw-runtime-resources-precache.js` File

Reviewer 2: Not sure if this is the most accurate title. Feel free to rephrase if needed, but avoid using code formatting and file names in headings.
```suggestion
=== Ignore Service Worker Resources Precache File
```

---
title: Overview
order: 0
layout: page
---

:vaadin-version: {moduleMavenVersion:com.vaadin:vaadin}
pass:[<!-- vale Vaadin.Versions = NO -->]

= Upgrading From Vaadin 15 To {vaadin-version}

> Upgrading from a Vaadin 15+ version to the latest is in general straightforward, nevertheless, a few breaking changes need to take into account when upgrading:

Reviewer 3: Maybe "[...] breaking changes need to *be* taken into account [...]" or "[...] breaking changes to take into account [...]". The colon at the end looks misplaced , should end with period.

> Upgrading from a Vaadin 15+ version to the latest is in general straightforward, nevertheless, a few breaking changes need to take into account when upgrading:

Reviewer: Updated suggestion based on latest version below.

[role="since:com.vaadin:vaadin@V19"]
== Generated @Id Field Is Now Optional In TypeScript

> A field with `@Id` annotation in Java is now optional in the generated TypeScript code. 

Reviewer 3: From the first reading understood this as the field being optional, i.e., may not be generated. Perhaps "[...] is now of optional type [...]" to clarify.

Given an entity with an `id` field:

.Entity.java
[source,java]
----
public class Entity {
    @Id
    private int id;
}
----

Now in the TypeScript files, instead of using `endpoint.getEntity(entity.id)`, you might need to change to `endpoint.getEntity(entity.id!)` or `endpoint.getEntity(entity.id?)`.

== Spring Security

> The Spring Security configuration requires some changes for the service worker related files.

Reviewer 3: "[...] to the service worker related files".

.SecurityConfiguration.java
[source,java]
----
@Override
protected void configure(HttpSecurity http) throws Exception {
    ...
    http.authorizeRequests().anyRequest().hasAnyAuthority(Role.getAllRoles());
    ...
}
----

In such case you would have configured Spring Security to bypass the static resources as below:

.SecurityConfiguration.java
[source,java]
----
@Override
public void configure(WebSecurity web) {
    web.ignoring().antMatchers(
            // client-side JS code
            "/VAADIN/**",
            ...
            // web application manifest
            "/manifest.webmanifest",
            "/sw.js",
            "/offline-page.html",
            ...
    );
}
----

You need to add one more file `/sw-runtime-resources-precache.js` to the list:

.SecurityConfiguration.java
[source,java]
----
@Override
public void configure(WebSecurity web) {
    web.ignoring().antMatchers(
            // client-side JS code
            "/VAADIN/**",
            ...
            // web application manifest
            "/manifest.webmanifest",
            "/sw.js",
            "/offline-page.html",
            "/sw-runtime-resources-precache.js",
            ...
    );
}

> ----

Reviewer 3: It seems too verbose to have this code snippet twice. Can it be done with only the latter version.

[role="since:com.vaadin:vaadin@V19"]
=== Ignore the Service Worker Initiated Requests
Another potential Spring Security related breaking change is about using `HttpSecurity.requestCache()` to redirect the user to the intended page after login. 

An example of using `HttpSecurity.requestCache()`:

.SecurityConfiguration.java
[source,java]
----
@Override
protected void configure(HttpSecurity http) throws Exception {
    ...
    http

    // Register our CustomRequestCache, that saves unauthorized access attempts, so
    // the user is redirected after login.
    .requestCache().requestCache(new CustomRequestCache())
    
    // Restrict access to our application.
    .and().authorizeRequests()

    // Allow all flow internal requests.
    .requestMatchers(SecurityUtils::isFrameworkInternalRequest).permitAll()
    ...
}
----

> Now you need to ignore the service worker initiated requests, otherwise the access attempts will be overridden by the service worker requests and Spring cannot redirect you to the intended page. 

Reviewer 3: Could be good to add a sentence here that this can be done by inspecting the `Referer` header as an explanation to the below code.
